### DISCUSSION TITLE
How to suppress copying of certain directories under the `static` directory for certain locales while building the site

### DISCUSSION THREAD

**Discussion by allenscha** [facebook/docusaurus#9722] [Q&A] [answered]
We maintain documentation for a PaaS product that supports multiple platforms, and we are keeping API reference files generated from platform-specific tools such as Dokka, Jazzy, Doxygen, and Typedoc as static resources. While the documentation itself supports multiple locales, the API reference files are in English only and located under the `static` directory, as follows:

```
static/api-reference/*
```

Since the API reference is in English only, we use `autoAddBaseUrl: false` in the sidebar link config so that the link to the API reference is independent of the current locale.

```
    type: 'link',
    label: 'API Reference',
    href: 'pathname:///api-reference/*',
    autoAddBaseUrl: false,
```

However, we recently found out that the content of `static/api-reference/*` is copied to all locales during the build process, as follows:

```
// build result
build/api-reference/*
build/{locale1}/api-reference/*
build/{locale2}/api-reference/*
...
```

Among the files above, all the files in the locale-specific directories are unused and only take up storage space.

Is there a way to improve the current situation? (For your reference, the tested version of Docusaurus is 2.2.0.)

**Comment by slorber** [comment]
That's the behavior we choose for the static directory, and it makes sense for many other use-cases to keep it that way.

If you don't want one copy of the api ref per locale, then you can simply not use the static directory.

You could build your site and at the end do `cp -R my-generated-api-ref website/build/api-reference`.
The static folder basically only do that copy/paste for you (really, nothing more), except it does that for each locale.

You don't need Docusaurus in practice to "merge" static files of multiple tools, and `cp -R` is enough. Just make sure that static files do not conflict with each others to prevent unwanted overrides. Using a baseUrl is generally a good idea it prevent that, instead of having multiple static sites to be hosted at the site root.

**Comment by allenscha** [comment]
> If you don't want one copy of the api ref per locale, then you can simply not use the static directory.
> 
> You could build your site and at the end do `cp -R my-generated-api-ref website/build/api-reference`. The static folder basically only do that copy/paste for you (really, nothing more), except it does that for each locale.

There's other considerations that prevent us from using this suggestion. We need to verify the API reference as part of the site URL while previewing the site with `yarn start`, and using the static directory and accessing the files through `pathname:///api-reference/*` seems to be the only option we have for that purpose.

If we keep using the static directory, a workaround would be to add a script to delete the API reference files copied to each locale directory and make the script run after `yarn build` in the deployment process, but that seems a bit unproductive.

So I'm wondering if it's feasible to add a configuration to exclude some directory under `static` from being copied to other local directories during build.

**Reply by slorber** [reply, marked as answer]
> but that seems a bit unproductive.

Hey that solves your use-case and is easy so why isn't this good enough?

I feel like we are trying to achieve useless micro-optimisations here. 

Is this really a problem in practice to deploy this duplicate api ref? 

Is copying for each locale + deleting a perf issue of any kind that's worth us to fix?

---

There is not config to exclude static directories, but we have `staticDirectories` to let you provide your own list.

https://docusaurus.io/docs/3.0.1/api/docusaurus-config#staticDirectories

Combined with the locale env variable (https://github.com/facebook/docusaurus/issues/4542#issuecomment-1434839071), that could solve the same use-case.

You could for example only list an api-ref static directory if locale is undefined or english.

**Comment by allenscha** [comment]
Another thing to note is that, in order to use localized assets (images, diagrams, etc.) for each locale in a multi-language site, a typical method would be to place asset files by locale under `static` as follows:

```
static
  /{locale1}
  /{locale2}
  /{locale3}
...
```

or 

```
static
  /img
    /{locale1}
    /{locale2}
    /{locale3}
...
```

and all these files would be copied to each locale directory during build, while most of them would be unused except in the corresponding locale.

This might be a more complicated case than the original inquiry, but I think it would be beneficial if there is some configuration to control the inclusion/exclusion of certain files under `static` for build.

**Reply by slorber** [reply]
This config should work:

```js
export default {
  staticDirectories: [
    'static',
    'staticLocalized/' + (process.env.DOCUSAURUS_CURRENT_LOCALE ?? "en"),
  ],
};
```

Details in https://github.com/facebook/docusaurus/issues/4542#issuecomment-1434839071

**Comment by charIeszhao** [comment]
Hi, same issue here. Is there an option to disable copying static files to different locales when building, I would prefer using the same static folder for all my i18n locales. Thanks.

**Reply by slorber** [reply]
No, we don't have that option. Doing so would break Docusaurus when deploying i18n sites on multiple subdomains.

But you don't have to use the Docusaurus `/static` folder, it's only there by convention and for convenience.

What you can do instead is put your static files in a different folder and copy it into your `website/build` folder as part of the build process

```bash
cp -R my-shared-static-folder website/build
```

This is basically the only thing that this Docusaurus feature is doing.

Then when referencing static files, don't use the baseUrl helpers or the Link component but a regular `<a>` tag that won't append baseUrl to it.

**Reply by charIeszhao** [reply]
Well, thanks for your reply. I tried the workaround above with the `staticLocalized`, and it worked for me. Thanks.

This is so important as it helps remove redundant images from my static folder. Previously the localized images are copied to every locale. Now it works perfectly for our docs site: https://docs.logto.io
